catalog/rocks/pgrefs.go

```diff
@@ -0,0 +1,222 @@
+			repositoryID, repository.StorageNamespace, repository.CreationDate, repository.DefaultBranchID)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(`
+				INSERT INTO kv_branches (repository_id, id, staging_token, commit_id)
+				VALUES ($1, $2, $3, $4)`,
+			repositoryID, repository.DefaultBranchID, branch.stagingToken, branch.CommitID)
+		return nil, err
+	}, db.WithContext(ctx))
+	return err
+}
+
+func (m *PGRefManager) ListRepositories(ctx context.Context, from RepositoryID) (RepositoryIterator, error) {
+	return NewRepositoryIterator(ctx, m.db, IteratorPrefetchSize, string(from)), nil
+}
+
+func (m *PGRefManager) DeleteRepository(ctx context.Context, repositoryID RepositoryID) error {
+	_, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		var err error
+		_, err = tx.Exec(`DELETE FROM kv_branches WHERE repository_id = $1`, repositoryID)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(`DELETE FROM kv_commits WHERE repository_id = $1`, repositoryID)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(`DELETE FROM kv_repositories WHERE id = $1`, repositoryID)
+		return nil, err
+	}, db.WithContext(ctx))
+	return err
+}
+
+func (m *PGRefManager) RevParse(ctx context.Context, repositoryID RepositoryID, ref Ref) (Reference, error) {
+	return ResolveRef(ctx, m, repositoryID, ref)
+}
+
+func (m *PGRefManager) GetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) (*Branch, error) {
+	branch, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		pbranch := &PgBranch{}
+		err := tx.Get(pbranch,
+			`SELECT staging_token, commit_id FROM kv_branches WHERE repository_id = $1 AND id = $2`,
+			repositoryID, branchID)
+		if err != nil {
+			return nil, err
+		}
+		return &Branch{
+			CommitID:     pbranch.CommitID,
+			stagingToken: pbranch.StagingToken,
+		}, nil
+	}, db.ReadOnly(), db.WithContext(ctx))
+	if errors.Is(err, db.ErrNotFound) {
+		return nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	return branch.(*Branch), nil
+}
+
+func (m *PGRefManager) SetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, branch Branch) error {
+	_, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		_, err := tx.Exec(`
+			INSERT INTO kv_branches (repository_id, id, staging_token, commit_id)
+			VALUES ($1, $2, $3, $4)
+				ON CONFLICT (repository_id, id)
+				DO UPDATE SET staging_token = $3, commit_id = $4`,
+			repositoryID, branchID, branch.stagingToken, branch.CommitID)
+		return nil, err
+	}, db.WithContext(ctx))
+	return err
+}
+
+func (m *PGRefManager) DeleteBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error {
+	_, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		_, err := tx.Exec(
+			`DELETE FROM kv_branches WHERE repository_id = $1 AND id = $2`,
+			repositoryID, branchID)
+		return nil, err
```

Comment:
Should it return an error if branch does not exist?

Reply:
this makes deletes idempotent and allows for blind writes. but I'm fine either way.